cpp/modmesh/buffer/BufferExpander.hpp

```diff
@@ -0,0 +1,173 @@
+    {
+        return BufferExpander::construct(copy_concrete(), /*clone*/ false);
+    }
+
+    BufferExpander(std::shared_ptr<ConcreteBuffer> const & buf, bool clone, ctor_passkey const &)
+        : m_concrete_buffer(clone ? buf->clone() : buf)
+        , m_begin(m_concrete_buffer->data())
+        , m_end(m_begin + size())
+        , m_end_cap(m_begin + m_concrete_buffer->size())
+    {
+    }
+
+    BufferExpander(size_type nbyte, ctor_passkey const &)
+    {
+        expand(nbyte);
+    }
+
+    BufferExpander(ctor_passkey const &) {}
+
+    BufferExpander() = delete;
+    BufferExpander(BufferExpander const &) = delete;
+    BufferExpander(BufferExpander &&) = delete;
+    BufferExpander & operator=(BufferExpander const &) = delete;
+    BufferExpander & operator=(BufferExpander &&) = delete;
+    ~BufferExpander() = default;
+
+    size_type size() const noexcept
+    {
+        return static_cast<size_type>(this->m_end - this->m_begin);
+    }
+
+    size_type capacity() const noexcept
+    {
+        return static_cast<size_type>(this->m_end_cap - this->m_begin);
+    }
+
+    void reserve(size_type cap);
+
+    void expand(size_type length)
+    {
+        reserve(length);
+        m_end = m_begin + length;
+    }
+
+    std::shared_ptr<ConcreteBuffer> copy_concrete(size_type cap = 0);
+    std::shared_ptr<ConcreteBuffer> const & as_concrete(size_type cap = 0);
+    bool is_concrete() const { return bool(m_concrete_buffer); }
+
+    int8_t operator[](size_t it) const { return data(it); }
+    int8_t & operator[](size_t it) { return data(it); }
+
+    int8_t at(size_t it) const
+    {
+        validate_range(it);
+        return data(it);
+    }
+    int8_t & at(size_t it)
+    {
+        validate_range(it);
+        return data(it);
+    }
+
+    /* Backdoor */
+    int8_t data(size_type it) const { return data()[it]; }
+    int8_t & data(size_type it) { return data()[it]; }
+    int8_t const * data() const noexcept { return data<int8_t>(); }
+    int8_t * data() noexcept { return data<int8_t>(); }
+    // clang-format off
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    template <typename T> T const * data() const noexcept { return reinterpret_cast<T *>(m_begin); }
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    template <typename T> T * data() noexcept { return reinterpret_cast<T *>(m_begin); }
+    // clang-format on
+
+private:
+
+    void validate_range(size_type it) const
+    {
+        if (it >= size())
+        {
+            throw std::out_of_range(Formatter() << "BufferExpander: index " << it << " is out of bounds with size " << size());
+        }
+    }
+
+    static std::unique_ptr<int8_t> allocate(size_type nbytes)
+    {
+        std::unique_ptr<int8_t> ret(nullptr);
+        if (0 != nbytes)
+        {
+            ret = std::unique_ptr<int8_t>(new int8_t[nbytes]);
+        }
+        return ret;
+    }
+
+    std::unique_ptr<int8_t> m_data_holder = nullptr;
```

Comment:
`m_data_holder` may be use more frequently than `m_concrete_buffer` because this class focuses on managing memory for collecting data.

Reply:
question, why not reuse `using unique_ptr_type = std::unique_ptr<int8_t, data_deleter_type>;` in ConcreteBuffer.hpp?

Reply:
`using unique_ptr_type = std::unique_ptr<int8_t, data_deleter_type>;` is not used in `BufferExpander` because I do not want the deleter yet.  We may need to write sophisticated reallocation code for the buffer managed by `m_data_holder`, and a deleter complicates it.

Maybe eventually we will need a deleter, but let's wait for a while.  The deleter in `ConcreteBuffer` wasn't added until it was needed.  And in the case of `BufferExpander`, maybe we will never need a deleter.  I hope we minimize the chance of overengineering.